prql/src/semantic/scope.rs

```diff
@@ -0,0 +1,226 @@
+use anyhow::Result;
+use enum_as_inner::EnumAsInner;
+use serde::Deserialize;
+use serde::Serialize;
+use std::collections::HashMap;
+use strum_macros::Display;
+
+use crate::ast::*;
+use crate::error::Span;
+
+/// Scope within which we can reference variables, functions and tables
+/// Provides fast lookups for different names.
+#[derive(Debug, Default, Serialize, Deserialize, Clone)]
+pub struct Context {
+    // current table columns (result of last pipeline)
```

Comment:
Great comments! We can make these into doc strings!

```suggestion
    /// current table columns (result of last pipeline)
```